doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,601 @@
+expected::
+
+    np.array([100], dtype=np.uint8) + 1000 == np.array([1100], dtype=np.uint16)
+
+But the following will then be a surprise::
+
+    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)
+
+Considering this, we believe that the proposal follows the "principle of least
+surprise".
+
+
+Usage and Impact
+================
+
+There will be no transition period due to the difficulty and noise this is
+expected to create.  In rare cases users may need to adjust code to avoid
+reduced precision or incorrect results.
+
+We plan to provide an *optional* warning mode capable of notifying users of
+potential changes in behavior in most relevant cases.
+
+
+Impact on ``can_cast``
+----------------------
+
+Can cast will never inspect the value anymore.  So that the following results
+are expected to change from ``True`` to ``False``::
+
+  np.can_cast(100, np.uint8)
+  np.can_cast(np.int64(100), np.uint8)
+  np.can_cast(np.array(100, dtype=np.int64), np.uint8)
+
+We expect that the impact of this change will be small compared to that of
+the following changes.
+
+.. note::
+
+    The first example where the input is a Python scalar could be preserved
+    to some degree, but this is not currently planned.
+
+
+Impact on operators and functions involving NumPy arrays or scalars
+-------------------------------------------------------------------
+
+The main impact on operations not involving Python scalars (float, int, complex)
+will be that 0-D arrays and NumPy scalars will never behave value-sensitive.
+This removes currently surprising cases.  For example::
+
+  np.arange(10, dtype=np.uint8) + np.int64(1)
+  # and:
+  np.add(np.arange(10, dtype=np.uint8), np.int64(1))
+
+Will return an int64 array because the type of ``np.int64(1)`` is strictly
+honoured.
+
+
+Impact on operators involving Python ``int``, ``float``, and ``complex``
+------------------------------------------------------------------------
+
+This NEP attempts to preserve the convenience that the old behaviour
+gave when working with literal values.
+The current value-based logic had some nice properties when "untyped",
+literal Python scalars are involved::
+
+  np.arange(10, dtype=np.int8) + 1  # returns an int8 array
+  np.array([1., 2.], dtype=np.float32) * 3.5  # returns a float32 array
+
+But led to complexity when it came to "unrepresentable" values:
+
+  np.arange(10, dtype=np.int8) + 256  # returns int16
+  np.array([1., 2.], dtype=np.float32) * 1e200  # returns float64
+
+The proposal is to preserve this behaviour for the most part.  This is achieved
+by considering Python ``int``, ``float``, and ``complex`` to be "weakly" typed
+in these operations.
+Hoewver, to mitigate user surprises, we plan to make conversion to the new type
+more strict:  This means that the results will be unchanged in the first
+two examples.  For the second one, the results will be the following::
+
+  np.arange(10, dtype=np.int8) + 256  # raises a TypeError
```

Comment:
Agreed with @rossbar. This is likely to break a ton of essentially harmless code. I think it would be great to raise on integer overflow (consistently, not just with python scalars), but I think it would be vastly preferable to only do that when we actually fall out of (u)int64.

It's laudable to be more precise on this (i.e. if someone _really_ wants to make sure their `int8` stays an `int8`), but I think it's going to be a very sharp edge for the large majority of users who do not care about the underlying representation of their ints, but will be hit with errors much more often.